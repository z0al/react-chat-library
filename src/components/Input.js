// @ts-check

// Packages
import React from 'react';
import resize from 'autosize';

/**
 * @typedef Props
 * @property {string} [text]	Default text
 * @property {string} [placeholder]	Input placeholder
 * @property {boolean} [submitOnEnter]	Trigger submit event on enter key
 * @property {number} [lines]	Number of visible lines without scrolling
 * @property {object} [containerStyle]
 * @property {object} [inputStyle]
 * @property {function} [onSubmit]
 * @property {function} [onTextChange]
 * @param {Props} props
 */
const Input = props => (
	<div style={props.containerStyle}>
		<textarea
			style={props.inputStyle}
			data-testid={'text-input'}
			wrap="soft"
			rows={1}
			value={props.text}
			placeholder={props.placeholder}
			onChange={e => {
				const input = e.target;
				props.onTextChange(input.value);

				// Resize based on content
				resize(input);
			}}
			onKeyDown={e => {
				// Enter key pressed without shift combination
				if (props.submitOnEnter && e.keyCode === 13 && !e.shiftKey) {
					e.preventDefault();
					props.onSubmit();
				}
			}}
		/>
		<style jsx>{`
			div {
				display: flex;
				flex: 1;
				background: white;
				box-sizing: border-box;
				border: 1px solid gray;
				margin: 5px 10px;
				padding: 0.3em 0.6em;
				border-radius: 5px;
			}

			textarea {
				padding: 0;
				margin: 0;
				flex-grow: 1;
				resize: none;
				border: none;
				outline: none;
				max-height: ${props.lines * 1.2}em;
			}
		`}</style>
	</div>
);

/** @type Props */
Input.defaultProps = {
	text: '',
	placeholder: 'Type a message',
	submitOnEnter: true,
	lines: 5,
	containerStyle: {},
	inputStyle: {},
	onSubmit: () => {},
	onTextChange: () => {}
};

export default Input;
