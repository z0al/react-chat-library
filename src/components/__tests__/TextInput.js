import React from 'react';
import renderer from 'react-test-renderer';
import { render } from 'react-testing-library';

import Input from '../Input';

test('should match the snapshot', () => {
	const tree = renderer.create(<Input />).toJSON();
	expect(tree).toMatchSnapshot();
});

test('should set initial text from Context', () => {
	const text = 'Hello world';
	const { getByTestId } = render(<Input text={text} />);
	expect(getByTestId('text-input').innerHTML).toEqual(text);
});
