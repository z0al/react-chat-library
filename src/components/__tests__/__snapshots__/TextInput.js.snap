// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match the snapshot 1`] = `
<div
  className="jsx-2974810480"
  style={Object {}}
>
  <textarea
    className="jsx-2974810480"
    data-testid="text-input"
    defaultValue=""
    onChange={[Function]}
    onKeyDown={[Function]}
    placeholder="Type a message"
    rows={1}
    style={Object {}}
    wrap="soft"
  />
</div>
`;
